File: 0547-number-of-provinces/0547-number-of-provinces.py
class UnionFind:
    def __init__(self, n):
        self.parent = [i for i in range(n)]
        self.rank = [1] * n
    
    def find(self, x):
        if (self.parent[x] != x):
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    
    def union(self, x, y):
        xset = self.find(x)
        yset = self.find(y)
        if (xset == yset):
            return
        if (self.rank[xset] < self.rank[yset]):
            self.parent[xset] = yset
        elif (self.rank[xset] > self.rank[yset]):
            self.parent[yset] = xset
        else:
            self.parent[yset] = xset
            self.rank[xset] += 1

class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        dsu = UnionFind(n)
        findCircleNum = n
        for i in range(n):
            for j in range(i + 1, n):
                if (isConnected[i][j] == 1 and dsu.find(i) != dsu.find(j)):
                    findCircleNum -= 1
                    dsu.union(i, j)
        return findCircleNum

File: 0547-number-of-provinces/0547-number-of-provinces.java
class UnionFind {
    private int[] parent;
    private int[] rank;
    public UnionFind(int size) {
        parent = new int[size];
        for (int i = 0; i < size; i++) {
            parent[i] = i;
        }
        rank = new int[size];
    }
    
    public int find(int x) {
        if (parent[x] != x) {
            return find(parent[x]);
        }
        return parent[x];
    }
    
    public void union(int x, int y) {
        int xset = find(x);
        int yset = find(y);
        if (xset == yset) return;
        if (rank[xset] < rank[yset]) parent[xset] = yset;
        else if (rank[xset] > rank[yset]) parent[yset] = xset;
        else {
            parent[yset] = xset;
            rank[xset]++;
        }
    }
}
class Solution {
    public int findCircleNum(int[][] isConnected) {
        int n = isConnected.length;
        UnionFind dsu = new UnionFind(n);
        int nFindCircleNum = n;
        for (int i = 0; i < n; i++)
            for (int j = i + 1; j < n; j++) {
                if (isConnected[i][j] == 1 && dsu.find(i) != dsu.find(j)) {
                    nFindCircleNum--;
                    dsu.union(i, j);
                }
            }
        return nFindCircleNum;
    }
}